### ISSUE TITLE
IllegalState error on Receiver::attach() with the home page client sample code

### ISSUE THREAD

**Issue by olivier-fs** [minghuaw/fe2o3-amqp#122]
Hi!

I'm using fe2o3-amqp-0.7.0, with tokio-1.21.2
I'm running RabbitMQ 3.11.1 with Erlang 25.1.1 on Windows 10.

I'm getting an error "IllegalState" when trying to receive data.
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IllegalState', fe2o3-amqp-test1-client\src\main.rs:34:10
```

I'm using the client sample https://github.com/minghuaw/fe2o3-amqp#client.

[ FYI I had to add
```
use fe2o3_amqp::types::messaging::Outcome;
```
to the build th succeed , but it's not related to my runtime error. ]

The backtrace is :
```
$ RUST_BACKTRACE=1 cargo run -p fe2o3-amqp-test1-client
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target\debug\fe2o3-amqp-test1-client.exe`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IllegalState', fe2o3-amqp-test1-client\src\main.rs:26:13
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library\std\src\panicking.rs:584
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library\core\src\panicking.rs:142
   2: core::result::unwrap_failed
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library\core\src\result.rs:1814
   3: enum$<core::result::Result<fe2o3_amqp::link::receiver::Receiver,enum$<fe2o3_amqp::link::error::ReceiverAttachError, 0, 5, RemoteClosedWithError>

 > >::unwrap<fe2o3_amqp::link::receiver::Receiver,enum$<fe2o3_amqp::link::error::ReceiverAttachError, 0, 5, Rem
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\core\src\result.rs:1107
   4: fe2o3_amqp_test1_client::main::async_block$0
             at .\fe2o3-amqp-test1-client\src\main.rs:22
   5: core::future::from_generator::impl$1::poll<enum$<fe2o3_amqp_test1_client::main::async_block_env$0> >
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\core\src\future\mod.rs:91
   6: core::future::future::impl$1::poll<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_test1_client::main::async_block_env$0> > >

>
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\core\src\future\future.rs:124
   7: tokio::runtime::scheduler::current_thread::impl$9::block_on::closure$0::closure$0::closure$0<core::pin::Pin<ref_mut$<core::future::from_generato

r::GenFuture<enum$<fe2o3_amqp_test1_client::main::async_block_env$0> > > > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:525
   8: tokio::coop::with_budget::closure$0<enum$<core::task::poll::Poll<tuple$<> > >,tokio::runtime::scheduler::current_thread::impl$9::block_on::closu

re$0::closure$0::closure_env$0<core::pin::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2o3_amqp
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\coop.rs:102
   9: std::thread::local::LocalKey<core::cell::Cell<tokio::coop::Budget> >::try_with<core::cell::Cell<tokio::coop::Budget>,tokio::coop::with_budget::c

losure_env$0<enum$<core::task::poll::Poll<tuple$<> > >,tokio::runtime::scheduler::current_thread::impl$9::block
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\std\src\thread\local.rs:445
  10: std::thread::local::LocalKey<core::cell::Cell<tokio::coop::Budget> >::with<core::cell::Cell<tokio::coop::Budget>,tokio::coop::with_budget::closu

re_env$0<enum$<core::task::poll::Poll<tuple$<> > >,tokio::runtime::scheduler::current_thread::impl$9::block_on:
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\std\src\thread\local.rs:421
  11: tokio::coop::with_budget
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\coop.rs:95
  12: tokio::coop::budget
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\coop.rs:72
  13: tokio::runtime::scheduler::current_thread::impl$9::block_on::closure$0::closure$0<core::pin::Pin<ref_mut$<core::future::from_generator::GenFutur

e<enum$<fe2o3_amqp_test1_client::main::async_block_env$0> > > > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:525
  14: tokio::runtime::scheduler::current_thread::Context::enter<enum$<core::task::poll::Poll<tuple$<> > >,tokio::runtime::scheduler::current_thread::i

mpl$9::block_on::closure$0::closure_env$0<core::pin::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:349
  15: tokio::runtime::scheduler::current_thread::impl$9::block_on::closure$0<core::pin::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2

o3_amqp_test1_client::main::async_block_env$0> > > > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:524
  16: tokio::runtime::scheduler::current_thread::impl$9::enter::closure$0<tokio::runtime::scheduler::current_thread::impl$9::block_on::closure_env$0<c

ore::pin::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_test1_client::main::async_block
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:595
  17: tokio::macros::scoped_tls::ScopedKey<tokio::runtime::scheduler::current_thread::Context>::set<tokio::runtime::scheduler::current_thread::Context

,tokio::runtime::scheduler::current_thread::impl$9::enter::closure_env$0<tokio::runtime::scheduler::current_thr
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\macros\scoped_tls.rs:61
  18: tokio::runtime::scheduler::current_thread::CoreGuard::enter<tokio::runtime::scheduler::current_thread::impl$9::block_on::closure_env$0<core::pin

::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_test1_client::main::async_block_env$0>
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:595
  19: tokio::runtime::scheduler::current_thread::CoreGuard::block_on<core::pin::Pin<ref_mut$<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_

test1_client::main::async_block_env$0> > > > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:515
  20: tokio::runtime::scheduler::current_thread::CurrentThread::block_on<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_test1_client::main::

async_block_env$0> > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\scheduler\current_thread.rs:161
  21: tokio::runtime::Runtime::block_on<core::future::from_generator::GenFuture<enum$<fe2o3_amqp_test1_client::main::async_block_env$0> > >
             at C:\Users\olivier\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.21.2\src\runtime\mod.rs:490
  22: fe2o3_amqp_test1_client::main
             at .\fe2o3-amqp-test1-client\src\main.rs:44
  23: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52\library\core\src\ops\function.rs:248
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error: process didn't exit successfully: `target\debug\fe2o3-amqp-test1-client.exe` (exit code: 101)
```

I also tried with the "rt" tokio runtime instead of "rt-multi-thread", and
```
#[tokio::main(flavor="current_thread")]
```
but I get the same error.

In the backtrace we can see
```
fe2o3_amqp::link::error::ReceiverAttachError, 0, 5, RemoteClosedWithError>
```

but the rabbit log says the client TCP connection was unexpectedly closed :
```
2022-10-16 16:19:03.390000+02:00 [info] <0.2201.0> accepting AMQP connection <0.2201.0> ([::1]:59147 -> [::1]:5672)
2022-10-16 16:19:03.392000+02:00 [debug] <0.2201.0> User 'guest' authenticated successfully by backend rabbit_auth_backend_internal
2022-10-16 16:19:03.392000+02:00 [info] <0.2201.0> AMQP 1.0 connection <0.2201.0>: user 'guest' authenticated
2022-10-16 16:19:03.393000+02:00 [debug] <0.2201.0> AMQP 1.0 connection.open frame: hostname = localhost, extracted vhost = /, idle_timeout = 60000
2022-10-16 16:19:03.396000+02:00 [debug] <0.2211.0> User 'guest' authenticated successfully by backend rabbit_auth_backend_internal
2022-10-16 16:19:03.398000+02:00 [info] <0.2201.0> AMQP 1.0 connection <0.2201.0>: user 'guest' authenticated and granted access to vhost '/'
2022-10-16 16:19:10.376000+02:00 [warning] <0.2201.0> closing AMQP connection <0.2201.0> ([::1]:59147 -> [::1]:5672):
2022-10-16 16:19:10.376000+02:00 [warning] <0.2201.0> client unexpectedly closed TCP connection
2022-10-16 16:19:10.377000+02:00 [debug] <0.2245.0> Closing all channels from connection '[::1]:59147 -> [::1]:5672' because it has been closed
2022-10-16 16:19:10.377000+02:00 [debug] <0.2247.0> Closing all channels from connection '[::1]:59147 -> [::1]:5672' because it has been closed
```
So it seems both sides consider the other has (unexpectedly) closed the connection.

I'm using default "guest" creds, and we can see there is no problem with authentification 

If I reorder the sample code to do the complete sending of data before attaching the receiver, the data is actually sent.

If I run a python client code (using pika-1.3.0 lib) with same creds/queue, the python client can receive the data sent by the rust code.

The IllegalState error is weird because it occurs even if the recever's attach() is actually the first operation performed on the session (i.e. when I remove the sender code)

Changing link / queue names gives same result : read success with pika, IllegalState with fe2o3.

**Comment by minghuaw** [comment]
Thanks for reporting the issue. I have not tested with a rabbit broker yet. I will set up a test environment and look into it.

**Comment by minghuaw** [comment]
@olivier-fs I have found the problem. It was my misunderstanding of the spec, though this misunderstanding seem common among other AMQP 1.0 brokers (eg. ActiveMQ, artemis, etc.)

The original spec says

> If there is no pre-existing terminus, and the peer does not wish to create a new one, is indicated by setting the local terminus (source or target as appropriate) to null

I have the sender and receiver checking both the `source` and `target` upon receiving a remote Attach. Most implementations that I have tested (ActiveMQ, ActiveMQ-Artemis, Service Bus, Event Hubs, qpid-dispatch) choose to reply with an empty but not null `source` or `target`. However, in the case of rabbitmq, it chooses to reply with an Attach with the `target` field set to null, which technically is correct.

This is a quick fix (already done locally), but I will need to test with other brokers for compatibility. I think you can expect to see a bug fix release in one or two hours.

Plus, there is actually one bug with rabbitmq's implementation, not a significant one though. The spec says

> A peer closes a link by sending the [detach](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-detach) frame with the handle for the specified link, and the closed flag set to true. The partner will destroy the corresponding link endpoint, and reply with its own [detach](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-detach) frame with the closed flag set to true.

The problem is that even if you send a closing detach, rabbit doesn't reply with a closing detach, which is indicated as an error though the close process is still completed. the easiest workaround is then to use a non-closing detach by

```rust
sender.detach().await.unwrap();
receiver.detach().await.unwrap();
```

**Comment by olivier-fs** [comment]
You were fast !
Yeah implementation specific differences, differnet ways of reading the spec, and obviously often enough, errors.
Even with RabbitMQ and the Erlang guys, so at least we Rust & C++ guys are not the only one to sometimes have bugs in our code ;-)
I will continue to make tests with your next release and RabbitMQ (I'm actually testing rabbit vs nats vs nng).
Thank you, nice work !

**Comment by minghuaw** [comment]
Compatibilty test:

- [x] rabbitmq with AMQP 1.0 plugin
- [x] activemq-artemis
- [x] qpid-dispatch-router
- [x] activemq (classic)
- [x] servicebus
- [x] event hubs

**Comment by minghuaw** [comment]
@olivier-fs "0.7.1" is published. Feel free to close the issue if you it fixes the bug :)

**Comment by olivier-fs** [comment]
I can confirm v0.7.1 fixed the issue with Receiver::attach() when using RabbitMQ.

FYI I hit the same 2nd issue you told about the close() operation (either sender or receiver btw).
Using close() with unwrap() I get
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DetachedByRemote'
```
But any serious production code should not unwrap and rather match the Err() and filter DetachedByRemote which is not an actual error.
(I very often have to do the same thing in C++/sockets/ASIO to filter 'connection reset by peer' "errors").
Not sure though if close() should silently discard the error....  May be the subject for a separate issue.
And I'm fine with detach() : it purposely consumes self, so I conclude I do not have to call close() after.

Deserves the checkmark for RabbitMQ on the compatibility matrix !

Even if I haven't done tests with heavy loads, I'll close this issue since all my recv ops are now successful.

**Comment by minghuaw** [comment]
Thanks for the quick tests. 

The `.close()` and `.detach()` API are provided only to differentiate a closing detach vs. a non-closing detach. A closing detach simply means the link cannot be resumed, and everything else is the same as a non-closing detach, so using `.detach()` is still correct.

The `DetachedByRemote` error is really meant for detecting remote peer actively detaching, and current design of link actually completes the whole detaching process upon receiving a remote detach and return this error. Actively closing a link shares the same detaching process, but it is because rabbit isn't really complying with the spec on replying to a closing detach that causes this error. 

Feel free to let me know of any question or problem with the crate. I'm always happy to help :)
